training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala

```diff
@@ -0,0 +1,467 @@
+            case "hinge" => computeHingeGradient(score, fv, options)
+            case "regression" => computeRegressionGradient(score, fv, options)
+            case _ => computeHingeGradient(score, fv, options)
+          }
+          // back-propagation for updating gradient
+          // note: activations have been computed in forwardPropagation
+          val outputLayerId = model.getNumHiddenLayers
+          val func = model.getActivationFunction.get(outputLayerId)
+          // delta: gradient of loss function w.r.t. node input
+          // activation: the output of a node
+          val outputNodeDelta = computeActivationGradient(score, func) * grad
+
+          backPropagation(model, outputNodeDelta.toFloat, gradient, flatFeatures, options.weightDecay.toFloat)
+        })
+        gradient.iterator
+      })
+      .mapValues(fv => (fv, 1.0))
+      .reduceByKey((a, b) => {
+        a._1.add(b._1)
+        (a._1, a._2 + b._2)
+      })
+      .mapValues(x => {
+        x._1.scale(1.0f / x._2.toFloat)
+        x._1
+      })
+      .collectAsMap
+      .toMap
+  }
+
+  def backPropagation(model: MlpModel,
+                      outputNodeDelta: Float,
+                      gradient: scala.collection.mutable.HashMap[(String, String), FloatVector],
+                      flatFeatures: java.util.Map[String, java.util.Map[java.lang.String, java.lang.Double]],
+                      weightDecay: Float = 0.0f) = {
+    // outputNodeDelta: gradient of the loss function w.r.t the input of the output node
+    val numHiddenLayers = model.getNumHiddenLayers
+    val layerNodeNumber = model.getLayerNodeNumber
+    val activationFunctions = model.getActivationFunction
+    // set delta for the output layer
+    var upperLayerDelta = new FloatVector(1)
+    upperLayerDelta.set(0, outputNodeDelta)
+    // compute gradient for bias at the output node
+    val outputBiasGrad = new FloatVector(1)
+    outputBiasGrad.set(0, outputNodeDelta)
+    val outputBiasKey = (LAYER_PREFIX + numHiddenLayers.toString, BIAS_PREFIX)
+    outputBiasGrad.add(gradient.getOrElse(outputBiasKey, new FloatVector(1)))
+    gradient.put(outputBiasKey, outputBiasGrad)
+    // update for hidden layers
+    for (i <- (0 until numHiddenLayers).reverse) {
+      // i decreases from numHiddenLayers-1 to 0
+      val numNode = layerNodeNumber.get(i)
+      val numNodeUpperLayer = layerNodeNumber.get(i + 1)
+      val func = activationFunctions.get(i)
+      val thisLayerDelta = new FloatVector(numNode)
+      // compute gradient of weights from the i-th layer to the (i+1)-th layer
+      val activations = model.getLayerActivations.get(i)
+      val hiddenLayerWeights = model.getHiddenLayerWeights.get(i)
+      val biasKey = (LAYER_PREFIX + i.toString, BIAS_PREFIX)
+      val biasGrad = gradient.getOrElse(biasKey, new FloatVector(numNode))
+      for (j <- 0 until numNode) {
+        val key = (LAYER_PREFIX + i.toString, NODE_PREFIX + j.toString)
```

Comment:
the drop out part is implemented in the model, because the activations corresponding to dropped nodes are zeros, the gradient for the weights that connecting that activation with the next node would be 0, so we don' t need to take special care here. will add some comments